addComponentsAndDirectDependencies

klu2 opened the pull request:
The idea behind this new method is to provide a convenience method to show all components of a container in the ComponentView as well as all ingoing and outgoing elements (persons, software systems, containers, components from other containers).

That way I get a view of my container who it is influencing and whom it is influenced by.

Relationships between between elements outside my container are removed, that means, if I have 3 containers:

A -> B
A -> C
B -> C

when I have a look at the ContainerView of A, then I want to see B and C with the relationships to A, but I do not want to see the relationship between B and C there

see https://structurizr.com/workspace/111#2 and https://structurizr.com/workspace/111#3 as examples of that code

simonbrowndotje commented:
Having used Structurizr to model some real-world software systems, I rarely find that I add all components inside a container to a component view (it's usually too much information).

Could we split out the "addAllComponents()" part of that method and do something like this instead? https://github.com/structurizr/java/compare/klu2-componentView-addDirectDependencies

klu2 commented:
agree.

I merged your branch into the pull request and fixed the javadoc.

klu2 commented:
I added another commit right now after applying this new method to my model.

It turns out that now it can easily happen that components of container A are included in the ComponentView of container B and this is wrong imho, because in the ComponentView of B it then looks like the component of A is part of B.

I therefore added a check that only "own" components might be added to the ComponentView of a container, never others.

And that required me to add the ugly addElement method in ComponentView again; I added some javadoc there

klu2 commented:
I did another refactoring now, making the code a bit cleaner; what I'm doing now in case I find relationships from components of other containers then I add that container to the view.

what would still be missing then probably is the relationship from that container to my inside elements, that is then related to your new addImplicitRelationship (which would add all of those implicit relationships though, so I won't use that here)

klu2 commented:
I added some more logging now to make that visible to the user, seems good now to me

simonbrowndotje commented:
I just made a few minor tweaks to my branch, and added a test case, which I'm not sure should pass ... it demonstrates that you end up with some orphaned elements when there are dependencies with the container but not with components inside it.

klu2 commented:
I merged your branch with the test in my branch.

basically you're right, as there are dependencies to and from the component

but the problem with ComponentViews is that on the web page you show the container as a dashed box (having its components inside) and afaik there is no possibility to draw an arrow to or from that dashed box. relationships that are visible need to go to the components inside.

simonbrowndotje commented:
Indeed, that's why it's called a component view. :-) You're right, relationships will need to go to a component in order to be visible. With that in mind, I think this line can be removed -> https://github.com/structurizr/java/blob/klu2-componentView-addDirectDependencies/structurizr-core/src/com/structurizr/view/ComponentView.java#L158

(which makes insideElements really just a set of components)

klu2 commented:
it could be removed as it will not add additional relationships to the view, but it may add other elements to the view, i.e. all elements that use my container. 

if they use the container then they most probably use one of the components inside (i.e. an API, a webservice, or whatever), so most probably that means that there is a missing relation ship to a component inside.

klu2 commented:
hm but yes, it is a **Component**View so probably you're right, and by calling addDirectDependencies we should only collect all dependencies to the components inside, and not to the container itself...

klu2 commented:
I updated that now such that references to/from the Container itself are not included anymore

simonbrowndotje commented:
I made a few tweaks and also renamed the method to be addExternalDependencies, which I think summarises that we're trying to do. Thoughts?

klu2 commented:
merged your branch into mine; ready for merging into master from my point of view
